src/job_market.rs

```diff
@@ -0,0 +1,154 @@
+use parking_lot::{Condvar, Mutex};
+use std::{collections::VecDeque, sync::Arc};
+
+/// A market for synchronising the sharing of jobs.
+///
+/// Maintains synchronisation for multiple threads, including shutdown behaviour once one finishes
+/// or panics.
+pub struct JobMarket<Job> {
```

Comment:
Not a blocker: perhaps this becomes `pub(crate)` to ensure the type doesn't get accidentally published.